Proposta de melhorias - Validações NFe

marcosgerene opened the pull request:
Esse PR é uma proposta de melhoria e não foi devidamente testado, então não deve ser aceito sem uma bateria de testes.

A ideia aqui é fazer algumas melhorias pontuais nos serviços do NFe:
 - Salvar o xml ANTES da validação, permitindo que, em caso de não validação seja possível para o desenvolvedor ver o arquivo com falhas;
 - Configuração para validar ou não os schemas, permitindo que o Schemas seja ignorado (o comportamento padrão é mantido por default);
 - Melhoria no validador exibindo todas as falhas do schemas e não mais uma a uma como acontece hoje

marcosgerene committed:
Merge pull request #63 from ZeusAutomacao/master

Atualizando repositório

marcosgerene committed:
[*] Salvar XML do envio do evento antes do validar, permitindo checar o motivo da não validação

marcosgerene committed:
[+] Configuração permitir ignorar validação de Schemas

marcosgerene committed:
[Danfe] Impressão do Danfe Simplificado

marcosgerene committed:
Merge pull request #64 from ZeusAutomacao/master

Atualizando reposositório

danilobreda commented on Shared.NFe.Utils/Validacao/Validador.cs:
```diff
@@ -127,8 +128,12 @@ public static void Valida(ServicoNFe servicoNFe, VersaoServico versaoServico, st
             Valida(servicoNFe, versaoServico, stringXml, loteNfe, pathSchema);
         }
 
+        private static StringBuilder Falhas { get; set; }
```
por ser static, na web pode acontecer de ter vários acessos na mesma propriedade em requisições diferentes, precisa de um lock talvez... Vou verificar melhor isso.

danilobreda commented on Shared.NFe.Utils/Validacao/Validador.cs:
```diff
@@ -156,21 +161,21 @@ public static void Valida(ServicoNFe servicoNFe, VersaoServico versaoServico, st
             }
+
+            if (Falhas.Length > 0)
+                throw new Exception($"Ocorreu o seguinte erro durante a validação XML: {Environment.NewLine}{Falhas}");
```
O correto não é retornar o ValidacaoSchemaException ?

Também seria interessante adicionar uma Lista<string> no ValidacaoSchemaException dos erros que teve e retornar eles também. No meu caso por exemplo, eu retorno uma lista de string com cada erro para o frontend, não retorno tudo em uma só string, pois o front formata da melhorar maneira.

danilobreda reviewed (commented):
ta ai alguns problemas que encontrei, o negócio do static realmente é um problemão! preciso testar aqui.

marcosgerene committed:
[*] Melhoras no validador
- Removendo StringBuilder estatico
- Gerando o exception do tipo ValidacaoSchemaException
- Função passa a retornar um array de string com as falhas (caso existam)

marcosgerene committed:
Merge branch 'master' of https://github.com/marcosgerene/DFe.NET
